pkg/source/source.go

```diff
@@ -67,34 +67,33 @@ type SyncingSource interface {
 	WaitForSync(ctx context.Context) error
 }
 
-// NewKindWithCache creates a Source without InjectCache, so that it is assured that the given cache is used
-// and not overwritten. It can be used to watch objects in a different cluster by passing the cache
-// from that other cluster.
-func NewKindWithCache(object client.Object, cache cache.Cache) SyncingSource {
-	return &kindWithCache{kind: Kind{Type: object, cache: cache}}
+// Kind creates a KindSource with the given cache provider.
+func Kind(cache cache.Cache, object client.Object) SyncingSource {
+	return &kind{obj: object, cache: cache}
 }
 
-type kindWithCache struct {
-	kind Kind
-}
-
-func (ks *kindWithCache) Start(ctx context.Context, handler handler.EventHandler, queue workqueue.RateLimitingInterface,
-	prct ...predicate.Predicate) error {
-	return ks.kind.Start(ctx, handler, queue, prct...)
-}
-
-func (ks *kindWithCache) String() string {
-	return ks.kind.String()
-}
+// KindAsPartialMetadata modifies the given Source created with Kind(...)
+// to project the type with a metav1.PartialObjectMetadata.
+func KindAsPartialMetadata(src SyncingSource, scheme *runtime.Scheme) error {
+	ks, ok := src.(*kind)
+	if !ok {
+		return fmt.Errorf("expected Source created with source.Kind(...), got %T", src)
+	}
 
-func (ks *kindWithCache) WaitForSync(ctx context.Context) error {
-	return ks.kind.WaitForSync(ctx)
+	metaObj := &metav1.PartialObjectMetadata{}
+	gvk, err := apiutil.GVKForObject(ks.obj, scheme)
+	if err != nil {
+		return fmt.Errorf("unable to determine GVK of %T for a metadata-only watch: %w", ks.obj, err)
+	}
+	metaObj.SetGroupVersionKind(gvk)
+	ks.obj = metaObj
+	return nil
 }
 
-// Kind is used to provide a source of events originating inside the cluster from Watches (e.g. Pod Create).
-type Kind struct {
-	// Type is the type of object to watch.  e.g. &v1.Pod{}
-	Type client.Object
+// kind is used to provide a source of events originating inside the cluster from Watches (e.g. Pod Create).
+type kind struct {
+	// obj is the type of object to watch.  e.g. &v1.Pod{}
+	obj client.Object
```

Comment:
Should this be objType? (just wondering why obj here vs ownerType elsewhere, fells like we could keep it consistent)